--- a/grammar.js
+++ b/grammar.js
@@ -27,4 +27,6 @@
   conflicts: ($) => [
+    [$._modified_type, $.type_constructor_pattern],
+    [$.named_type, $.type_constructor_pattern],
     [$._expression, $._postfix_expression_base],
     [$.qualified_operator, $._postfix_expression_base],
     [$._expression, $.qualified_operator, $._postfix_expression_base],
@@ -570,7 +572,7 @@
       seq(
         optional(field("visibility", $.visibility_modifier)),
         field("name", $.identifier),
-        optional(seq("::", field("value", $.integer_literal))),
+        optional(seq("::", optional("-"), field("value", $.integer_literal))),
         ";",
       ),
 
@@ -2485,6 +2487,8 @@
 
     _primary_expression: ($) =>
       choice(
+        $.array_type,
+        $.many_pointer_type,
         $.integer_literal,
         $.float_literal,
         $.char_literal,
@@ -2755,7 +2759,7 @@
     prefix_operator: (_) =>
       token(prec(1, choice("!!", "!", "?", "-", "+", "~", /\*+/, "&"))),
 
-    suffix_operator: (_) => token(prec(2, "++")),
+    suffix_operator: (_) => token(prec(2, choice("++", /!{2,}/))),
 
     range_operator: (_) => token(prec(2, choice("..=", ".."))),
 
