--- a/grammar.js
+++ b/grammar.js
@@ -777,7 +777,6 @@
         "#no_context",
         "#dump",
         "#must",
-        "#noreturn",
         "#returns_twice",
         $.inline_directive,
       ),
@@ -792,7 +791,6 @@
         "#no_context",
         "#dump",
         "#must",
-        "#noreturn",
         "#returns_twice",
         $.inline_directive,
       ),
@@ -1084,6 +1082,7 @@
         $.for_statement,
         $.inline_statement,
         $.try_defer_statement,
+        $.on_drop_error_statement,
         $.defer_statement,
         $.caller_defer_statement,
         $.using_statement,
@@ -1293,6 +1292,9 @@
 
     inline_statement: ($) =>
       seq("#inline", field("statement", choice($.while_statement, $.for_statement))),
+
+    on_drop_error_statement: ($) =>
+      seq("#on_drop_error", "(", field("owner", $.identifier), ",", field("error", $.identifier), ")", field("body", $.block)),
 
     try_defer_statement: ($) =>
       seq("#try", "defer", field("statement", $._statement)),
@@ -1478,6 +1480,7 @@
         $.for_statement,
         $.inline_statement,
         $.try_defer_statement,
+        $.on_drop_error_statement,
         $.defer_statement,
         $.caller_defer_statement,
         $.using_statement,
@@ -1706,6 +1709,7 @@
         $.generic_type_variable,
         $.context_type,
         $.self_type,
+        $.never_type,
         $.string_literal,
         $.named_type,
       ),
@@ -1749,6 +1753,7 @@
         $.generic_type_variable,
         $.context_type,
         $.self_type,
+        $.never_type,
         $.string_literal,
         $.named_type,
       ),
@@ -1899,6 +1904,8 @@
     context_type: (_) => "#Context",
 
     self_type: (_) => "#Self_Type",
+
+    never_type: (_) => "#Never",
 
     self_expression: (_) => "#self",
 
@@ -2155,6 +2162,7 @@
         $.context_expression,
         $.self_expression,
         $.self_type,
+        $.never_type,
         $.source_location_expression,
         $.code_expression,
         $.lambda_expression,
@@ -2480,6 +2488,7 @@
         $.context_expression,
         $.self_expression,
         $.self_type,
+        $.never_type,
         $.source_location_expression,
         $.code_expression,
         $.lambda_expression,
@@ -2736,7 +2745,7 @@
     prefix_operator: (_) =>
       token(prec(1, choice("!!", "!", "?", "-", "+", "~", /\*+/, "&"))),
 
-    suffix_operator: (_) => token(prec(2, "++")),
+    suffix_operator: (_) => token(prec(2, choice("++", /!{2,}/))),
 
     range_operator: (_) => token(prec(2, choice("..=", ".."))),
 
