--- a/grammar.js
+++ b/grammar.js
@@ -27,2 +27,4 @@
   conflicts: ($) => [
+    [$._modified_type, $.type_constructor_pattern],
+    [$.named_type, $.type_constructor_pattern],
     [$._expression, $._postfix_expression_base],
@@ -572,3 +574,3 @@
         field("name", $.identifier),
-        optional(seq("::", field("value", $.integer_literal))),
+        optional(seq("::", optional("-"), field("value", $.integer_literal))),
         ";",
@@ -2165,2 +2167,4 @@
         $.string_literal,
+        $.cstring_literal,
+        $.byte_array_literal,
         $.string_block,
@@ -2354,2 +2358,4 @@
         $.string_literal,
+        $.cstring_literal,
+        $.byte_array_literal,
         $.boolean_literal,
@@ -2418,3 +2424,3 @@
           seq(field("object", $._postfix_expression_base), ".", "acast"),
-          seq(field("object", $._postfix_expression_base), ".*"),
+          seq(field("object", $._postfix_expression_base), choice(".*", ".&")),
           seq(
@@ -2487,2 +2493,4 @@
       choice(
+        $.array_type,
+        $.many_pointer_type,
         $.integer_literal,
@@ -2491,2 +2499,4 @@
         $.string_literal,
+        $.cstring_literal,
+        $.byte_array_literal,
         $.string_block,
@@ -2755,5 +2765,5 @@
     prefix_operator: (_) =>
-      token(prec(1, choice("!!", "!", "?", "-", "+", "~", /\*+/, "&"))),
-
-    suffix_operator: (_) => token(prec(2, "++")),
+      token(prec(1, choice("!!", "!", "?", "-", "+", "~", /\*+/, /&+/))),
+
+    suffix_operator: (_) => token(prec(2, choice("++", /!{2,}/))),
 
@@ -2805,2 +2815,8 @@
 
+    cstring_literal: (_) =>
+      token(seq('c"', repeat(choice(/[^"\\\n]/, /\\./)), '"')),
+
+    byte_array_literal: (_) =>
+      token(seq('b"', repeat(choice(/[^"\\\n]/, /\\./)), '"')),
+
     multiline_string_line: (_) => token(seq("\\\\", /[^\n]*/)),
